tests.texlua/memoize-extract.lua

```diff
@@ -0,0 +1,1279 @@
+	local _io_lines = io.lines
+
+	---safely get an iterator over the lines of a file
+	---@param name string
+	---@return fun()? iterator
+	---@return string? error message
+	io_lines = function(name)
+		if kpse.in_name_ok_silent_extended(name) then
+			return _io_lines(name)
+		else
+			return nil, ("Opening (read) '%s' not permitted"):format(name)
+		end
+	end
+end
+
+-----------------------------------------
+-- security relevant functions go here --
+--       more complex functions        --
+-----------------------------------------
+
+-- restricted function defined here
+local extract_pages
+do
+	-- safe the functions/libraries needed in this restricted area
+	local lfs = lfs
+	-- this is not inside the function but the startup code -> can error here
+	if not lfs then error("lfs is not available. This script needs to be executed with texlua") end
+	local os_spawn = os.spawn
+	local os_rm    = os.remove
+
+	---extract all pages specified in `pages` from `src_pdf` to dedicated files specified via `out_prefix`
+	---can raise an error
+	---@param src_pdf string
+	---@param out_prefix string
+	---@param pages [integer]
+	---@param pdf_version string
+	---@return integer? return_code of the underlying os.execute
+	---@return string? error returned by os.execute
+	---@return function? cleanup clean up all files created in the process
+	---@return string? out_pat pattern to which the pages were written to
+	extract_pages = function(src_pdf, out_prefix, pages, pdf_version)
+		if not kpse.in_name_ok_silent_extended(src_pdf) then
+			return nil, ("Opening '%s' not permitted."):format(src_pdf), nil, nil
+		end
+
+		local out_pat = ("%s%%d.pdf.tmp"):format(escape_pattern_for_format(out_prefix))
+		if not kpse.out_name_ok_silent_extended(out_pat:format(0)) then
+			return nil, ("Writing to '%s' (and following) not permitted."):format(out_pat:format(0)), nil, nil
+		end
+
+		if not pdf_version:find("^%d%.%d$") then
+			return nil, ("Invalid pdf_version provided: %s"):format(pdf_version)
+		end
+
+		-- Be aware that using the %d syntax for -sOutputFile=... does not reflect the
+		-- page number in the original document. If you chose (for example) to process
+		-- even pages by using -sPageList=even, then the output of -sOutputFile=out%d.png
+		-- would still be out1.png, out2.png, out3.png etc.
+		local cmd = ([[rungs -dSAFER -sDEVICE=pdfwrite -dNOPAUSE -dQUIET -dBATCH -dAutoRotatePages=/None -dCompatibilityLevel="%s" -sPageList="%s" -sOutputFile="%s" "%s"]]):format(
+			pdf_version,
+			table.concat(pages, ","),
+			out_pat,
+			src_pdf
+		)
+		local succ, err = os_spawn(cmd)
+
+		-- removes generated files (in case they still exist)
+		local cleanup = function()
+			for i in ipairs(pages) do
+				local fn = out_pat:format(i)
+				-- this is cleanup -> fail silently, not throwing errors
+				if kpse.out_name_ok_silent_extended(fn) then
+					if lfs.isfile(fn) then
+						return os_rm(fn)
+					end
+				end
+			end
+		end
```

Comment:
It might be a good idea to check the file extensions here first. The TeX Live security model doesn't prevent documents from overwriting the main `.tex` file, but that would make users _very_ unhappy, so I'd recommend enforcing that the file extension is `.pdf`, `.tmp`, `.mmz`, or similar.

Reply:
Based on the program flow, it actually always ends with `.pdf.tmp` because of this check/construction:
```lua
		local out_pat = ("%s%%d.pdf.tmp"):format(escape_pattern_for_format(out_prefix))
		if not kpse.out_name_ok_silent_extended(out_pat:format(0)) then
			return nil, ("Writing to '%s' (and following) not permitted."):format(out_pat:format(0)), nil, nil
		end
```

But I agree, an `rm` is a quite destructive operation and we should be extra careful, so it's implemented in 39ab9bb